_posts/2017-11-26-setting-up-intellij-for-rust.md

```diff
@@ -0,0 +1,105 @@
+
+## Advanced Setup
+
+If you want your environment to do more, here are some of the tools I use to improve the work flow in rust and IntelliJ.
+
+### Format on save ([rustfmt](https://crates.io/crates/rustfmt/))
+
+1.  `cargo install rustfmt`
+2.  Install the "Save Actions" plugin in IntelliJ
+3.  Configure Save Actions like this
+    *   You want to activate save actions on shortcut - IntelliJ does a lot of autosaving which will result in your code being reformated while you are still typing it if you use activate on save
+    *   If you use the VIM emulator, `:wa` will not trigger the reformatting - you must use the save shortcut
+4.  (Optional) Rebind the default `ctrl+shift+s` shortcut to `ctrl+s` for more convenience
+
+### Clippy linting hinting ([clippy](https://crates.io/crates/clippy))
+
+![picture of clippy from microsoft word](http://images.dailytech.com/nimage/19706_Clippy3.jpg)
+
+Clippy is... clippy but for rust. It will make suggestions about improvements to be made to your code. The only difference between clippy and rust's clippy is that rust's clippy is actually useful.  
+
+1.  `rustup toolchain install nightly`
+2.  `rustup run nightly cargo install clippy`
+3.  cd to root directory of your project
+4.  `rustup run nightly cargo clippy`
+
+### Watch tasks ([cargo-watch](https://crates.io/crates/cargo-watch))
+
+When I'm writing code - I hate having to flip to the terminal to compile and run the tests and ensure it's all good. I like to use watch tasks to continuously run my tests. Rust has an option for this.  
+
+1.  `cargo install cargo-watch`
+2.  cd to root directory of your project
+3.  `cargo watch -x test`
+
+### Integrating watch and clippy into IntelliJ
+
+This is all well and good, but the watch tasks and clippy aren't integrated into IntelliJ, they're run separately on a terminal, and only one at a time! Well it's possible to bring them into your IDE so everything is in the same place and running together.  
+
+1.  Ensure the "Terminal" plugin is installed and enabled in IntelliJ
+2.  open the terminal in IntelliJ
+3.  `cargo watch -s "rustup run nightly cargo clippy && cargo test"`
+
+## Done!
+
+Now you have a rust environment capable of  
+
+* refactoring
+* syntax highlighting
+* code completion
+* code generation
+* error highlighting
+* linting suggestions
+* format on save
+* continuous testing
+
+You should be aware that IntelliJ does not support rust debugging. If that's important to you, consider using another IDE such as visual studio code, CLion, vim or emacs. If you are attached to IntelliJ then you can use `rust-lldb` (installed by default with rustup) from the terminal.
+## Questions ?
+Any questions ? Ask on https://users.rust-lang.org/ and ping me the link to the post on Twitter or email it to me at dan.cohen@codurance.com. This way the answer will be visible to everyone in the community.
```

Comment:
May help for the addresses to be clickable here

Reply:
not the email address - that always annoys me when it tries to open my 'default' email program